### PR TITLE
chapter_11

### PR DESCRIPTION
**Pull request by chuan97sun**

### CODE REVIEW

**Comment by chuan97sun** [comment]
ready

**Comment by chuan97sun** [comment]
ready

File: app/assets/stylesheets/account_activations.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the AccountActivations controller here.
+// They will automatically be included in application.css.
```
Comment (ngocvt-0484) [inline app/assets/stylesheets/account_activations.scss]: remove empty file you have not used

File: app/helpers/account_activations_helper.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+module AccountActivationsHelper
```
Comment (ngocvt-0484) [inline app/helpers/account_activations_helper.rb]: remove empty file you have not used

File: app/mailers/application_mailer.rb [hunk 3]
```diff
@@ -1,4 +1,4 @@
 class ApplicationMailer < ActionMailer::Base
-  default from: 'from@example.com'
+  default from: "noreply@example.com"
```
Comment (ngocvt-0484) [inline app/mailers/application_mailer.rb]: use ENV variable here
ENV["MAIL_FROM"] = "noreply@example.com"

File: app/mailers/user_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+class UserMailer < ApplicationMailer
+
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
```
Comment (ngocvt-0484) [inline app/mailers/user_mailer.rb]: remove all comment

File: app/mailers/user_mailer.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  #   en.user_mailer.account_activation.subject
+  #
+  def account_activation(user)
```
Comment (ngocvt-0484) [inline app/mailers/user_mailer.rb]: remove (

**Comment by chuan97sun** [comment]
ready

File: app/controllers/account_activations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+class AccountActivationsController < ApplicationController
+  def edit
+    user = User.find_by(email: params[:email])
```
Comment (saobv-2223) [inline app/controllers/account_activations_controller.rb]: remove ()

File: app/models/user.rb [hunk 7]
```diff
@@ -41,14 +40,30 @@ def forget
...
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false unless digest
+
+    BCrypt::Password.new(digest).is_password? token
+end
+
+  def activate
+   update_attribute :activated, true
+   update_attribute :activated_at, Time.zone.now
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
   update_attributes activated: true, activated_at: Time.zone.now
```

File: app/models/user.rb [hunk 8]
```diff
@@ -41,14 +40,30 @@ def forget
...
 
   private
   
   def downcase_email
     email.downcase!
   end
+
+  def create_activation_digest
+    self.activation_token = User.new_token
+    self.activation_digest = User.digest(activation_token)
```
Comment (saobv-2223) [inline app/models/user.rb]: remove ()

File: app/models/user.rb (hunk 7 above)
Comment (chuan97sun) [reply app/models/user.rb]: this oneline code cannot activate the user, i have tried just now

**Comment by chuan97sun** [comment]
ready

File: app/models/user.rb (hunk 7 above)
Comment (manhnt-0960) [reply app/models/user.rb]: please use: `update activated: true, activated_at: Time.zone.now`

File: config/routes.rb [hunk 9]
```diff
@@ -7,5 +7,6 @@
     get "/login", to: "sessions#new"
     post "/login", to: "sessions#create"
     delete "/logout", to: "sessions#destroy"
+    resources :account_activations, only: [:edit]
```
Comment (manhnt-0960) [inline config/routes.rb]:
```suggestion
    resources :account_activations, only: :edit
```

File: app/models/user.rb (hunk 7 above)
Comment (chuan97sun) [reply app/models/user.rb]: Mr Manh, it does not work also, even if I change to update_attributes,I cannot do the re-login of activated user if use this oneline code

**Comment by chuan97sun** [comment]
ready

File: app/helpers/sessions_helper.rb [hunk 10]
```diff
@@ -14,7 +14,7 @@ def current_user
       @current_user ||= User.find_by id: session[:user_id]
     elsif user_id = cookies.signed[:user_id]
       user = User.find_by id: user_id
-      if user&.authenticated? :remember, cookies[:remember_token]
+      if user && user.authenticated?(:remember, cookies[:remember_token])
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/mailers/application_mailer.rb [hunk 11]
```diff
@@ -1,4 +1,4 @@
 class ApplicationMailer < ActionMailer::Base
-  default from: 'from@example.com'
+  default from: ENV["env_email"]
   layout 'mailer'
```
Comment (saobv-2223) [inline app/mailers/application_mailer.rb]: using ", please

File: app/models/user.rb (hunk 7 above)
Comment (saobv-2223) [reply app/models/user.rb]: try to restart server ?

Comment (chuan97sun) [reply app/models/user.rb]: tried,still canoot,has re-login issue

File: app/mailers/application_mailer.rb (hunk 11 above)
Comment (chuan97sun) [reply app/mailers/application_mailer.rb:3]: edited but not pushed yet

File: app/helpers/sessions_helper.rb (hunk 10 above)
Comment (chuan97sun) [reply app/helpers/sessions_helper.rb]: remove the () and my whole app is crash, even after I added the () back and restart the server. the app cant be run, i even restarted my virtual machine to restart the app.

**Comment by chuan97sun** [comment]
ready

File: app/helpers/sessions_helper.rb (hunk 10 above)
Comment (chuan97sun) [reply app/helpers/sessions_helper.rb]: problem solved for this

File: config/environments/development.rb [hunk 12]
```diff
@@ -31,8 +31,20 @@
...
 
+  config.action_mailer.raise_delivery_errors = true
   # Don't care if the mailer can't send.
-  config.action_mailer.raise_delivery_errors = false
+  config.action_mailer.delivery_method = :smtp
+  host = "localhost"
+  config.action_mailer.default_url_options = { host: host, port: '3000', protocol: 'http' }
+  ActionMailer::Base.smtp_settings = {
+    address: 'smtp.gmail.com',
+    port: '587',
```
Comment (saobv-2223) [inline config/environments/development.rb]: using ", please

**Comment by chuan97sun** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**